aten/src/ATen/native/SpectralOps.cpp

```diff
@@ -257,4 +257,135 @@ Tensor stft(const Tensor& self, const int64_t n_fft, const optional<int64_t> hop
+       << ", window="; \
+    if (window.defined()) { \
+      SS << window.toString() << "{" << window.sizes() << "}"; \
+    } else { \
+      SS << "None"; \
+    } \
+    SS << ", center=" << center << ", normalized=" << normalized << ", onesided=" << onesided << ", length="; \
+    if (lengthOpt.has_value()) { \
+      SS << lengthOpt.value(); \
+    } else { \
+      SS << "None"; \
+    } \
+    SS << ")"
+
+  // default_init hop_length and win_length
+  const auto hop_length = hop_lengthOpt.value_or(n_fft >> 2);
+  const auto win_length = win_lengthOpt.value_or(n_fft);
+
+  const auto input_shape = self.sizes();
+  const auto input_dim = self.dim();
+  const auto n_frames = self.size(-2);
+  const auto fft_size = self.size(-3);
+
+  const auto expected_output_signal_len = n_fft + hop_length * (n_frames - 1);
+
+  const auto options = at::device(self.device()).dtype(self.dtype());
+  if (self.numel() == 0) {
+    std::ostringstream ss;
+    REPR(ss) << ": input tensor cannot be empty.";
+    AT_ERROR(ss.str());
+  }
+  if (input_dim != 4) {
+    std::ostringstream ss;
+    REPR(ss) << ": expected a tensor with 4 dimensions, but got " << input_dim;
+    AT_ERROR(ss.str());
+  }
+ if (self.size(-1) != 2) {
+    std::ostringstream ss;
+    REPR(ss) << ": expected the last dimension to be 2 (corresponding to real and imaginary parts), but got " << self.size(-1);
+    AT_ERROR(ss.str());
+  }
+
+  if (onesided) {
+    if (n_fft / 2 + 1 != fft_size) {
+      std::ostringstream ss;
+      REPR(ss) << ": expected the frequency dimension (3rd to the last) of the input tensor to match n_fft / 2 + 1 when onsided=True, but got " << fft_size;
+      AT_ERROR(ss.str());
+    }
+  } else {
+    if (n_fft != fft_size) {
+      std::ostringstream ss;
+      REPR(ss) << ": expected the frequency dimension (3rd to the last) of the input tensor to match n_fft when onsided=False, but got " << fft_size;
+      AT_ERROR(ss.str());
+    }
+  }
+
+  if (!(0 < hop_length && hop_length <= win_length)) {
+    std::ostringstream ss;
+    REPR(ss) << ": expected 0 < hop_length <= win_length";
+    AT_ERROR(ss.str());
+  }
+
+  if (!(0 < win_length && win_length <= n_fft)) {
+    std::ostringstream ss;
+    REPR(ss) << ": expected 0 < win_length <= n_fft";
+    AT_ERROR(ss.str());
+  }
+  if (window.defined()) {
+    if (window.dim() != 1 || window.size(0) != win_length) {
+      std::ostringstream ss;
+      REPR(ss) << ": Invalid window shape. window has to be 1D and length of `win_length`";
+      AT_ERROR(ss.str());
+    }
+  }
+
+  Tensor window_tmp = window.defined() ? window : at::ones({win_length,}, options);
+  if (win_length != n_fft) {
+    // center window by padding zeros on right and left side
+    int64_t left = (n_fft - win_length) / 2;
+    window_tmp = at::constant_pad_nd(window_tmp, {left, n_fft - win_length - left}, 0);
+    TORCH_INTERNAL_ASSERT(window_tmp.size(0) == n_fft);
```

Comment:
nit: As we discussed, the assertions do mimic the python implementation, but we could consider replacing some of them by `UserWarning` in a later iteration.